src/libmongoc/src/mongoc/mongoc-topology-private.h

```diff
@@ -62,14 +64,38 @@ typedef struct _mongoc_rr_data_t {
    char *txt_record_opts;
 } mongoc_rr_data_t;
 
+struct _mongoc_topology_t;
+
+MONGOC_DECL_SPECIAL_TS_POOL (
+   mongoc_server_session_t,
+   mongoc_server_session_pool,
+   struct _mongoc_topology_t,
+   /* ctor/dtor/prune are defined in the new_with_params call */
+   NULL,
+   NULL,
+   NULL)
+
 typedef bool (*_mongoc_rr_resolver_fn) (const char *service,
                                         mongoc_rr_type_t rr_type,
                                         mongoc_rr_data_t *rr_data,
                                         size_t initial_buffer_size,
                                         bson_error_t *error);
 
+typedef union mc_shared_tpl_descr {
+   mongoc_shared_ptr sptr;
+   mongoc_topology_description_t *ptr;
+} mc_shared_tpl_descr;
+
+static const mc_shared_tpl_descr MC_SHARED_TPL_DESCR_NULL = {
+   MONGOC_SHARED_PTR_NULL};
+
 typedef struct _mongoc_topology_t {
-   mongoc_topology_description_t description;
+   /**
+    * @brief The topology description. Do not access directly. Instead, use
+    * mc_tpld_take_ref()
+    */
+   mc_shared_tpl_descr _shared_descr_;
```

Comment:
I was thinking about it, but I then thought it would be better to keep it as strongly-typed as often as possible, since shared_ptr just contains a `void*`. I don't think it's a huge difference, and I'm still open to be convinced otherwise again.